File: PROMISES/free-lunch_PROMISES.py
import sys
graph = None

inf = 987654321

def update(v1, v2, cost, V):
    for i in xrange(V):
        for j in xrange(V):
            graph[i][j] = min(graph[i][j],\
                            min(graph[i][v1] + cost + graph[v2][j],\
                            graph[i][v2] + cost + graph[v1][j]))

def floyd(V):
    for i in xrange(V):
        for j in xrange(V):
            for k in xrange(V):
                graph[i][j] = min(graph[i][j], graph[i][k]+graph[k][j])

if __name__ == "__main__":
    rl = lambda : sys.stdin.readline()
    for _ in xrange(int(rl())):
        V, M, N = map(int, rl().split())
        V += 1
        graph = [[inf]*V for _ in xrange(V)]
        for i in xrange(V):
            graph[i][i] = 0

        for _ in xrange(M):
            v1, v2, cost = map(int, rl().split())
            graph[v1][v2] = cost
            graph[v2][v1] = cost

        count = 0
        for _ in xrange(N):
            v1, v2, cost = map(int, rl().split())
            if graph[v1][v2] <= cost:
                count += 1
                continue

            graph[v1][v2] = graph[v2][v1] = cost
            update(v1,v2,cost,V)

        print(count)

File: PROMISES/free-lunch_PROMISES.cpp
#include <iostream>
#include <vector>
#include <limits>
#include <algorithm>
#include <sstream>

using namespace std;

#define MAX 987654321
int V, adj[201][201];

void floyd(){
    for(int i = 0; i < V; i++ ){
        for(int j = 0; j < V; j++){
            for(int k = 0; k < V; k++){
                adj[i][j] = min(adj[i][j], adj[i][k]+adj[k][j]);
            }
        }
    }
}

void update(int v1, int v2, int cost){
    for(int i = 0; i < V; i++ ){
        for(int j = 0; j < V; j++){
                adj[i][j] = min(adj[i][j], \
                    min(adj[i][v1]+adj[v2][j]+cost,adj[i][v2]+adj[v1][j]+cost));
        }
    }
}

int main(){
    cin.sync_with_stdio(false);

    int T = 0;
    int M,N;

    cin >> T;
    while(T--){
        cin >> V >> M >> N;
        V++;
        for(int i = 0; i < V; i++)
            for(int j = 0; j < V; j++)
                adj[i][j] = MAX;
        for(int i = 0; i < V; i++)
            adj[i][i] = 0;

        int v1,v2,cost;
        for(int i= 0; i < M; i++){
            cin >> v1 >> v2 >> cost;
            adj[v1][v2] = cost;
            adj[v2][v1] = cost;
        }

        int count = 0;
        for(int i= 0; i < N; i++){
            cin >> v1 >> v2 >> cost;
            if(adj[v1][v2] <= cost) {
                count += 1;
                continue;
            }
            update(v1,v2,cost);
        }

        cout << count <<endl;
    }
    return 0;
}
